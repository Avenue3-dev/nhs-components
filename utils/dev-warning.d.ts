/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export declare const useExperimentalWarning: (component: 'tabs' | 'character-count') => void;
/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export declare const useDevWarning: (message: string) => void;
