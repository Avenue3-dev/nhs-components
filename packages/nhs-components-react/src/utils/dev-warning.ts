const prod = process.env.NODE_ENV === 'production'

export const experimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  if (prod) {
    return
  }

  console.warn(`ATTENTION
  This component is tagged as experimental and therefore could change without notice. See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}
