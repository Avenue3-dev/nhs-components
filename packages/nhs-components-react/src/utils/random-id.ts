export const generateRandomString = (length = 5): string => {
  const rand = Math.random() + 1

  return rand.toString(36).substring(2, length + 2)
}

export const generateRandomName = (prefix?: string): string => {
  const randString = generateRandomString()

  return prefix ? `${prefix}_${randString}` : randString
}

export const generateRandomId = generateRandomName
