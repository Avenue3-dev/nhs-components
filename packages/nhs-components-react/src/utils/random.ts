export const generateRandomString = (length: number = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
