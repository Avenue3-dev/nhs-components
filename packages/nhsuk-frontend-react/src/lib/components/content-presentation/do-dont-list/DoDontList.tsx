import { AsElementHeadingProps } from '@/types/heading'
import { HTMLProps, createContext, useContext } from 'react'
import clsx from 'clsx'
import { CrossIcon, TickIcon } from '@/icons'

export type DoDontListType = 'do' | 'dont'

type DoDontListItemProps = HTMLProps<HTMLLIElement>

type HeadingProps = AsElementHeadingProps & HTMLProps<HTMLHeadingElement>

const Heading: React.FC<HeadingProps> = ({
  children,
  className,
  asElement: Component = 'h3',
  ...rest
}): JSX.Element => {
  return (
    <Component
      className={clsx('nhsuk-do-dont-list__label', className)}
      {...rest}
    >
      {children}
    </Component>
  )
}

const Item: React.FC<DoDontListItemProps> = ({
  children,
  ...rest
}): JSX.Element => {
  const type = useContext(DoDontListContext)

  return (
    <li {...rest}>
      {type === 'do' ? <TickIcon /> : <CrossIcon />}
      {children}
    </li>
  )
}

const DoDontListContext = createContext<DoDontListType>('do')

type DoDontListProps = {
  type?: DoDontListType
  headingProps?: HeadingProps
} & HTMLProps<HTMLDivElement>

type DoDontList = {
  Item: typeof Item
} & React.FC<DoDontListProps>

/**
 * Use Do and Don't lists to help users understand more easily what they should and shouldn't do.
 *
 * For more information on when to use this component, go to the [NHS Service Manuals website](https://service-manual.nhs.uk/design-system/components/do-and-dont-lists).
 *
 * @link https://service-manual.nhs.uk/design-system/components/do-and-dont-lists
 *
 * @example
 * ```tsx
 * <DoDontList type="do">
 *  <DoDontList.Item>Do this</DoDontList.Item>
 *  <DoDontList.Item>Do that</DoDontList.Item>
 * </DoDontList>
 * ```
 */
const DoDontList: DoDontList = ({
  type = 'do',
  className,
  headingProps,
  children,
  ...rest
}) => {
  return (
    <div className={clsx('nhsuk-do-dont-list', className)} {...rest}>
      <Heading {...headingProps}>{type === 'do' ? 'Do' : "Don't"}</Heading>
      {/* eslint-disable-next-line jsx-a11y/no-redundant-roles */}
      <ul
        className={clsx('nhsuk-list', {
          'nhsuk-list--tick': type === 'do',
          'nhsuk-list--cross': type !== 'do',
        })}
        role="list"
      >
        <DoDontListContext.Provider value={type}>
          {children}
        </DoDontListContext.Provider>
      </ul>
    </div>
  )
}

DoDontListContext.displayName = 'DoDontListContext'
Item.displayName = 'DoDontList.Item'
DoDontList.displayName = 'DoDontList'

DoDontList.Item = Item

export default DoDontList
