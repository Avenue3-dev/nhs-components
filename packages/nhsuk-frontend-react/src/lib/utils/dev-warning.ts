const prod = process.env.NODE_ENV === 'production'

/**
 * Print a warning when using experimental components when not in production mode.
 *
 * @param component name of the experimental component.
 */
export const useExperimentalWarning = (
  component: 'tabs' | 'character-count',
): void => {
  useDevWarning(`This component is tagged as experimental and therefore could change without notice.

  See more details about this component at https://service-manual.nhs.uk/design-system/components/${component}
  `)
}

/**
 * Print a warning message when not in production mode.
 *
 * @param message message to print when not it production.
 */
export const useDevWarning = (message: string): void => {
  if (prod) {
    return
  }

  console.warn(message)
}
