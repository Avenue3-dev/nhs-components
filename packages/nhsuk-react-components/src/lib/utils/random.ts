export const generateRandomString = (length = 8): string => {
  return [...Array<string>(length)]
    .map(() => Math.random().toString(36)[2])
    .join('')
}
