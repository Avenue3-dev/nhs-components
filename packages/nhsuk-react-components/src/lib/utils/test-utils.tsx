/* eslint-disable import/export */
import {
  RenderOptions,
  RenderResult,
  cleanup,
  render,
} from '@testing-library/react'
import { MemoryRouter as Router } from 'react-router-dom'
import { afterEach } from 'vitest'

afterEach(() => {
  cleanup()
})

const customRender = (
  ui: React.ReactElement,
  options: RenderOptions = {},
): RenderResult => {
  return render(ui, {
    // wrap provider(s) here if needed
    wrapper: ({ children }) => <Router>{children}</Router>,
    ...options,
  })
}

export * from '@testing-library/react'
export { default as userEvent } from '@testing-library/user-event'
// override render export
export { customRender as render }
