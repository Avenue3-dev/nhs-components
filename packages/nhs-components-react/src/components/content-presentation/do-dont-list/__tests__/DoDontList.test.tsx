import { render, screen } from '@/utils/test-utils'
import { describe, expect, test } from 'vitest'
import DoDontList from '..'
import { DoDontListType } from '../DoDontList'

describe('DoDontList', () => {
  test('should match snapshot', () => {
    render(
      <DoDontList data-testid="do-dont-list">
        <DoDontList.Item>Do this</DoDontList.Item>
        <DoDontList.Item>Do that</DoDontList.Item>
      </DoDontList>,
    )

    expect(screen.getByTestId('do-dont-list')).toMatchSnapshot()
  })

  test.each(['do', 'dont'] as DoDontListType[])('should match snapshot - type %s', (type) => {
    render(
      <DoDontList data-testid="do-dont-list" type={type}>
        <DoDontList.Item>Do this</DoDontList.Item>
        <DoDontList.Item>Do that</DoDontList.Item>
      </DoDontList>,
    )

    expect(screen.getByTestId('do-dont-list')).toMatchSnapshot()
  })
})
