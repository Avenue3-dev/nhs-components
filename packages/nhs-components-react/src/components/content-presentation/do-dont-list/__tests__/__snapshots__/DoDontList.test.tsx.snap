// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DoDontList > should match snapshot - type do 1`] = `
<div
  class="nhsuk-do-dont-list"
  data-testid="do-dont-list"
>
  <h3
    class="nhsuk-do-dont-list__label"
  >
    Do
  </h3>
  <ul
    class="nhsuk-list nhsuk-list--tick"
    role="list"
  >
    <li>
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__tick"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.4 7.8l-8.5 8.4L5.6 12"
          stroke="#007f3b"
          stroke-linecap="round"
          stroke-width="4"
        />
      </svg>
      Do this
    </li>
    <li>
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__tick"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.4 7.8l-8.5 8.4L5.6 12"
          stroke="#007f3b"
          stroke-linecap="round"
          stroke-width="4"
        />
      </svg>
      Do that
    </li>
  </ul>
</div>
`;

exports[`DoDontList > should match snapshot - type dont 1`] = `
<div
  class="nhsuk-do-dont-list"
  data-testid="do-dont-list"
>
  <h3
    class="nhsuk-do-dont-list__label"
  >
    Don't
  </h3>
  <ul
    class="nhsuk-list nhsuk-list--cross"
    role="list"
  >
    <li>
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__cross"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
          fill="#d5281b"
        />
        <path
          d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
          fill="#d5281b"
        />
      </svg>
      Do this
    </li>
    <li>
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__cross"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M17 18.5c-.4 0-.8-.1-1.1-.4l-10-10c-.6-.6-.6-1.6 0-2.1.6-.6 1.5-.6 2.1 0l10 10c.6.6.6 1.5 0 2.1-.3.3-.6.4-1 .4z"
          fill="#d5281b"
        />
        <path
          d="M7 18.5c-.4 0-.8-.1-1.1-.4-.6-.6-.6-1.5 0-2.1l10-10c.6-.6 1.5-.6 2.1 0 .6.6.6 1.5 0 2.1l-10 10c-.3.3-.6.4-1 .4z"
          fill="#d5281b"
        />
      </svg>
      Do that
    </li>
  </ul>
</div>
`;

exports[`DoDontList > should match snapshot 1`] = `
<div
  class="nhsuk-do-dont-list"
  data-testid="do-dont-list"
>
  <h3
    class="nhsuk-do-dont-list__label"
  >
    Do
  </h3>
  <ul
    class="nhsuk-list nhsuk-list--tick"
    role="list"
  >
    <li>
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__tick"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.4 7.8l-8.5 8.4L5.6 12"
          stroke="#007f3b"
          stroke-linecap="round"
          stroke-width="4"
        />
      </svg>
      Do this
    </li>
    <li>
      <svg
        aria-hidden="true"
        class="nhsuk-icon nhsuk-icon__tick"
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18.4 7.8l-8.5 8.4L5.6 12"
          stroke="#007f3b"
          stroke-linecap="round"
          stroke-width="4"
        />
      </svg>
      Do that
    </li>
  </ul>
</div>
`;
